pallets/crowdfunding/src/lib.rs

```diff
@@ -0,0 +1,619 @@
+            origin: OriginFor<T>,
+            crowdfund_key: CrowdFundKey,
+        ) -> DispatchResultWithPostInfo {
+            T::AuthorityOrigin::ensure_origin(origin)?;
+            let _ = Self::do_approve(crowdfund_key)?;
+            Ok(().into())
+        }
+
+        // What happens if the threshold is not met and the the contribution round is over?
+        // TODO: New extrinsic to cancel a project and unreserve all assets.
+        // Also allow for a project to undergo another contribution round.
+        // Like in briefs
+
+	}
+
+#[pallet::hooks]
+impl  <T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+    fn on_initialize(n: BlockNumberFor<T>) -> Weight {
+        let mut weight: Weight = Default::default(); 
+        let crowdfund_keys: BoundedKeysPerRound<T> = RoundsExpiring::<T>::take(n);
+        weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+
+        // SAFETY: BoundedKeysPerRound must be sane as to not have overweight blocks.
+        crowdfund_keys.iter().for_each(|key| {
+            CrowdFundsInRound::<T>::remove(key, RoundType::ContributionRound);
+            weight.saturating_add(T::DbWeight::get().reads_writes(1, 1));
+        });
+        
+        weight
+    }
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn new_crowdfund(
+        who: T::AccountId,
+        agreement_hash: H256,
+        proposed_milestones: BoundedProposedMilestones<T>,
+        required_funds: BalanceOf<T>,
+        currency_id: common_types::CurrencyId,
+    ) -> Result<CrowdFundKey, DispatchError> {
+        // Check if identity is required
+        if <T as Config>::IsIdentityRequired::get() {
+            let _ = Self::ensure_identity_is_decent(&who)?;
+        }
+
+        let crowdfund_key = CrowdFundCount::<T>::get();
+
+        // Todo: Take storage deposit>
+
+        // For now we keep them as proposed milestones until the project is able to submit.
+        let crowdfund = CrowdFund {
+            agreement_hash,
+            milestones: proposed_milestones,
+            contributions: BTreeMap::new().try_into().expect("empty BTree is smaller than bound; qed"),
+            required_funds,
+            currency_id,
+            raised_funds: (0_u32).into(),
+            initiator: who.clone(),
+            created_on: <frame_system::Pallet<T>>::block_number(),
+            approved_for_funding: false,
+            cancelled: false,
+            is_converted: false,
+        };
+
+        // Add crowdfund to list
+        <CrowdFunds<T>>::insert(crowdfund_key, crowdfund);
+
+		let next_crowdfund_key = crowdfund_key.checked_add(1).ok_or(Error::<T>::Overflow)?;
+        CrowdFundCount::<T>::put(next_crowdfund_key);
+
+        Self::deposit_event(Event::CrowdFundCreated(
+            who,
+            agreement_hash,
+            crowdfund_key,
+            required_funds,
+            currency_id,
+        ));
+
+        Ok(crowdfund_key)
+    }
+
+    pub fn try_update_existing_crowdfund(
```

Comment:
this can never fair right? if so just call it `update_crowdfund`

Reply:
done